PrivateEndpoint & PrivateLinkService APIs update

bowen0106 opened the pull request:
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

bowen0106 committed:
Add two properties on Subnet for privateIp. Add two new APIs to support privateIp.

Minor modify on readme.md file.

AutorestCI commented:
# Automation for azure-sdk-for-java

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-js

AutorestCI commented:
# Automation for azure-sdk-for-ruby

bowen0106 committed:
Fix some issue found by Model Validation.

azuresdkci commented:
Can one of the admins verify this patch?

EvgenyAgafonchikov commented on specification/network/resource-manager/Microsoft.Network/stable/2019-04-01/checkPrivateLinkServiceVisibility.json:
```diff
@@ -0,0 +1,83 @@
+  "paths": {
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Network/locations/{location}/checkPrivateLinkServiceVisibility": {
+      "post": {
+        "operationId": "CheckPrivateLinkServiceVisibility",
```
Please check if this expected operation name (i.e. plain operation name instead of `Class_Operation` form).
Current value would place this operation directly under INetworkManagementClient like [CheckDnsNameAvailability](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/network/Microsoft.Azure.Management.Network/src/Generated/INetworkManagementClient.cs#L485)

bowen0106 replied:
Thanks for the suggestion, I was thinking that the cmdlet that under different path from private link service,  should be a separate operation id. But since both you and Anton is suggested, I have already changed the operation id to merge them into private endpoint. This make code looks more neat. Thanks.

EvgenyAgafonchikov reviewed (commented):
Left some comments.

I also ran validation tools on the API, could you please also take a look into below? Likely we can change this API (I believe we can while in preview).

There is a pair of warning related to missing scheme in response (inserted below). When I tested UpdatePrivateEndpointConnection previous time, it returned connection body if I remember correctly. Please re-check this fact and update with scheme value if so.

```
PUT operation request body schema should be the same as its 200 response schema, to allow reusing the same entity between GET and PUT. If the schema of the PUT request body is a superset of the GET response body, make sure you have a PATCH operation to make the resource updatable. Operation: 'PrivateLinkServices_UpdatePrivateEndpointConnection
200/201 Responses of long running operations must have a schema definition for return type.
```

number213 commented:
Given that two new APIs are related to PrivateLinkServices, maybe it would be better to add them to `privateLinkServices.json` instead of creating two new files?

bowen0106 committed:
Merge all apis relative to private end point / private link service to its json file.

bowen0106 committed:
Fix a typo name from design spec that casue GetAvailablePrivateEndpointType get null property.

number213 commented on specification/network/resource-manager/readme.md:
```diff
@@ -940,6 +938,12 @@ directive:
     from: applicationSecurityGroup.json
     reason: name, id and type properties are inherited from the upper level
+  - suppress: RequiredPropertiesMissingInResourceModel
+    from: autoApprovedPrivateLinkServices.json
```
Since there is no `autoApprovedPrivateLinkServices.json` file, you don't need this exception.
Same for `availablePrivateEndpointTypes.json` and `checkPrivateLinkServiceVisibility.json`

bowen0106 replied:
Fixed. Thanks.

bowen0106 committed:
Change AutoApprovedPrivateLinkService structure to fit with NRP side.

bowen0106 commented on specification/network/resource-manager/Microsoft.Network/stable/2019-04-01/autoApprovedPrivateLinkServices.json:
```diff
@@ -0,0 +1,93 @@
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/locations/{location}/autoApprovedPrivateLinkServices": {
+      "get": {
+        "operationId": "AutoApprovedPrivateLinkServices_Get",
```
Already fixed. Thanks.

bowen0106 committed:
Delete suppress for merged APIs in readme.md.

EvgenyAgafonchikov commented:
Could you please review my comment above (response schema related). Adding previously recorded traffic for more context (there is response body received from server):
```
DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
PUT

Absolute Uri:
https://management.azure.com/subscriptions/<ID>/resourceGroups/test-evg/providers/Microsoft.Network/privateLinkServices/test/privateEndpointConnections/test.
19b38612-9576-4036-88d2-dc34e0cb5b18?api-version=2019-04-01

Body:
{
  "properties": {
    "privateEndpoint": {
      "properties": {
        "privateLinkServiceConnections": [],
        "manualPrivateLinkServiceConnections": []
      },
      "id": "/subscriptions/<ID>/resourceGroups/test-evg/providers/Microsoft.Network/privateEndpoints/test",
      "tags": {}
    },
    "privateLinkServiceConnectionState": {
      "status": "Rejected",
      "description": "change"
    }
  },
  "name": "test.19b38612-9576-4036-88d2-dc34e0cb5b18",
  "id":
"/subscriptions/<ID>/resourceGroups/test-evg/providers/Microsoft.Network/privateLinkServices/test/privateEndpointConnections/test.19b38612-9576-4036-88d2-dc34e0cb5b18"
}

DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
OK

Body:
{
  "name": "test.19b38612-9576-4036-88d2-dc34e0cb5b18",
  "id":
"/subscriptions/<ID>/resourceGroups/test-evg/providers/Microsoft.Network/privateLinkServices/test/privateEndpointConnections/test.19b38612-9576-4036-88d2-dc34e0cb5b18",
  "etag": "W/\"0cd819b1-aa1c-4ccc-ae61-de4928d37de0\"",
  "properties": {
    "provisioningState": "Updating",
    "privateEndpoint": {
      "id": "/subscriptions/<ID>/resourceGroups/test-evg/providers/Microsoft.Network/privateEndpoints/test"
    },
    "privateLinkServiceConnectionState": {
      "status": "Rejected",
      "description": "change",
      "actionsRequired": ""
    }
  },
  "type": "Microsoft.Network/privateLinkServices/privateEndpointConnections"
}

```

number213 commented on specification/network/resource-manager/Microsoft.Network/stable/2019-04-01/privateEndpoint.json:
```diff
@@ -303,6 +303,87 @@
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Network/locations/{location}/availablePrivateEndpointTypes": {
+      "get": {
+        "operationId": "AvailablePrivateEndpointTypes_List",
```
SDK generator would use `operationId` to get class and method names (e.g. `AvailablePrivateEndpointTypes.List()`). Since this method is related to privateEndpoints, maybe it would be better to name it `PrivateEndpoints_ListAvailablePrivateEndpointTypes` or `PrivateEndpoints_ListAvailableTypes`

bowen0106 replied:
I agree with. Previously I was thinking the cmdlet is under different path from private endpoint /private link service, so they should be in different operation class id. I have already fix it, this make code looks more neat. Thanks.

number213 commented on specification/network/resource-manager/Microsoft.Network/stable/2019-04-01/privateEndpoint.json:
```diff
@@ -303,6 +303,87 @@
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/locations/{location}/availablePrivateEndpointTypes": {
+      "get": {
+        "operationId": "AvailableResourceGroupPrivateEndpointTypes_List",
```
Since this method is related to privateEndpoints, maybe it would be better to name it `PrivateEndpoints_ListAvailableResourceGroupPrivateEndpointTypes` or `PrivateEndpoints_ListAvailableResourceGroupTypes `

bowen0106 replied:
Fixed.

number213 commented on specification/network/resource-manager/Microsoft.Network/stable/2019-04-01/privateLinkService.json:
```diff
@@ -422,6 +422,90 @@
+        "tags": [
+          "PrivateLinkServices"
+        ],
+        "operationId": "CheckPrivateLinkServiceVisibility",
```
Since this method is related to privateLinkService, maybe it would be better to name it `PrivateLinkServices_CheckPrivateLinkServiceVisibility`

bowen0106 replied:
Fixed.

number213 commented on specification/network/resource-manager/Microsoft.Network/stable/2019-04-01/privateLinkService.json:
```diff
@@ -422,6 +422,90 @@
+        "tags": [
+          "PrivateLinkServices"
+        ],
+        "operationId": "AutoApprovedPrivateLinkServices_List",
```
Since this method is related to privateLinkService, maybe it would be better to name it `PrivateLinkServices_ListAutoApprovedPrivateLinkServices`

bowen0106 replied:
Fixed.

anton-evseev reviewed (commented):
Comments for both `2019-04-01` and `2019-06-01`

bowen0106 committed:
1. Change value to PrivateEndpointConnection when calling put PrivateEndpointConnect.
2. Change opertaion id for opertaions that relative to private endpoint and private link service.

bowen0106 committed:
Minor fix for model validation.

EvgenyAgafonchikov commented on specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/privateEndpoint.json:
```diff
@@ -306,7 +306,7 @@
     "/subscriptions/{subscriptionId}/providers/Microsoft.Network/locations/{location}/availablePrivateEndpointTypes": {
       "get": {
-        "operationId": "AvailablePrivateEndpointTypes_List",
+        "operationId": "PrivateEndpoints_ListAvailablePrivateEndpointTypes",
```
This kind of naming seems to trigger new warnings like this:
```
Per the Noun_Verb convention for Operation Ids, the noun 'PrivateEndpoints' should not appear after the underscore. Note: If you have already shipped an SDK on top of this spec, fixing this warning may introduce a breaking change.
```
Could you please consider naming operations w/o noun on the right side of underscore? E.g. `PrivateEndpoints_ListAvailableTypes` or something like this.

bowen0106 replied:
Hi Evgeny, the AvailablePrivateEndpointType is a single noun. It's a single class type, different from privateEndpoint. So let's consider to still use 'PrivateEndpoints_ListAvailablePrivateEndpointTypes'. Thank you.

number213 replied:
I haven't noticed this earlier, but `AvailablePrivateEndpointTypes.List` has already been published in Network SDK and is being used in Azure PowerShell. So changing it would be a **breaking change** and we wouldn't want that. Could you please revert this change?
Also, I think in this case we would also need to change  `PrivateEndpoints_ListPrivateEndpointAvailableTypesByResourceGroup` to `AvailablePrivateEndpointTypes_ListByResourceGroup` to keep consistent naming with `List`/`ListByResourceGroup` pair.

bowen0106 replied:
@number213 Thanks , that was my concern too. I have already revert it back and add _ListByResourceGroup.

bowen0106 committed:
Revert opertion id changes for AvailablePrivateEndpointType to previous name.

bowen0106 committed:
1. Change CheckPrivateLinkServiceVisibility API, the input now has a body content that contains a private link service alias string.
2. Change both CheckPrivateLinkServiceVisibility/AutoApprovedPrivateLinkServices API to support for with/without ResourceGroup parameter.

number213 commented:
@KrisBash could you please give feedback from ARM's side?

ravbhatnagar reviewed (commented):
looks fine, just one minor note.

ravbhatnagar commented:
Signing off from ARM side with one minor comment. Not blocking but good to fix

bowen0106 commented on specification/network/resource-manager/Microsoft.Network/stable/2019-04-01/privateEndpoint.json:
```diff
@@ -420,6 +501,44 @@
+          "type": "string",
+          "description": "Resource type."
+        },
+        "resourceName": {
```
@ravbhatnagar Thanks for reviewing. I agreed with you. The name sounds inaccuracy. In design doc, it is called 'ServiceName' and in the NRP code it's 'ResouceName' , both are not good, but in order to works with NRP side correctly , I changed the name accordingly. So I guess we should leave it unless RP have changed.

tjprescott commented:
What's up with the JS SDK failing?

weinong commented:
@tjprescott @number213 who can help look at the SDK issue?

number213 commented:
I think this is the step that was removed by @sergey-shandar (https://github.com/Azure/azure-rest-api-specs/pull/6391). My guess is that it's deprecated and Network's release branch is out of sync with `master`.

weinong commented:
@bowen0106 can you try rebase your branch?

sauryadas commented:
@shahabhijeet can you please with the go sdk generation? or point to someone who can help?

shahabhijeet commented:
@jhendrixMSFT will the right person who can help you with Go SDK related questions.

jhendrixMSFT commented:
The build errors are a known issue for network (not just Go is affected).  I'm following up with the owners of the swagger-to-sdk job but for now the failures can be ignored.

jhendrixMSFT commented:
@AutorestCI regenerate azure-sdk-for-go

lmazuel commented:
@AutorestCI regenerate azure-sdk-for-go

AutorestCI commented:
<details><summary>Encountered an unknown error</summary><p>

```python
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/azure_devtools/ci_tools/github_tools.py", line 33, in exception_to_github
    yield context
  File "/usr/local/lib/python3.6/dist-packages/azure_devtools/ci_tools/bot_framework.py", line 105, in manage_comment
    response = getattr(self.handler, orderstr)(webhook_data.issue, *split_text)
  File "/usr/local/lib/python3.6/dist-packages/swaggertosdk/restapi/restbot.py", line 54, in regenerate
    return self.rebuild(issue, repotag)
  File "/usr/local/lib/python3.6/dist-packages/swaggertosdk/restapi/restbot.py", line 36, in rebuild
    self.sdk_default_base
  File "/usr/local/lib/python3.6/dist-packages/swaggertosdk/restapi/github_handler.py", line 151, in rest_pr_management
    sdk_tag=sdk_tag
  File "/usr/local/lib/python3.6/dist-packages/swaggertosdk/SwaggerToSdkNewCLI.py", line 310, in generate_sdk_from_git_object
    sdk_repo.git.push('origin', branch_name, set_upstream=True)
  File "/usr/local/lib/python3.6/dist-packages/git/cmd.py", line 548, in <lambda>
    return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/git/cmd.py", line 1014, in _call_process
    return self.execute(call, **exec_kwargs)
  File "/usr/local/lib/python3.6/dist-packages/git/cmd.py", line 825, in execute
    raise GitCommandError(command, status, stderr_value, stdout_value)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(1)
  cmdline: git push --set-upstream origin restapi_auto_6332
  stderr: 'To https://github.com/Azure/azure-sdk-for-go.git
 ! [rejected]            restapi_auto_6332 -> restapi_auto_6332 (fetch first)
error: failed to push some refs to 'https://AutorestCI:58ab395c311d1bd75b3e1db1cc8adaf9acc42afe@github.com/Azure/azure-sdk-for-go.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.'

```

</p></details>

sauryadas commented:
@jhendrixMSFT who's the owner of the swagger to sdk job? we really need this PR to merge ASAP so we can start our work on AKS private clusters which is scheduled to go in preview in 6 weeks.

jhendrixMSFT commented:
This was resolved yesterday by @lmazuel.  Regardless you can still merge the PR.
